_posts/2020-09-06-bench-rust-kotlin-microservices.md

```diff
@@ -0,0 +1,448 @@
+- Ktor reaches ~5000 req/s in the 32--128 connections zone.
+- Actix hits more than double of that, ~11000 req/s for 128 and 256 concurrent connections.
+- After reaching their peaks, both Actix and Ktor become oversaturated,
+  showing slightly degraded performance.
+  This effect is more pronounced and happens earlier in the Ktor case.
+- The behaviour of Http4k in extreme connection counts is *opposite*,
+  reaching its peak *successful* req/s of ~3500, despite the presence of non-zero error ratio.
+  - I am unable to reasonably explain this behaviour.
+
+Coincidence or not,
+Actix & Ktor that share somewhat similar shapes in the graph are both based on async executors and low thread counts,
+while Http4k uses more conventional high thread count and blocking I/O.
+
+Note that Google Cloud Run's default (and maximum) concurrent connection count of 80
+almost exactly hits the sweet spot of all measured implementations.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/latency_vs_connections_50.svg" />
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/latency_vs_connections_90.svg" />
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/latency_vs_connections_99.svg" />
+
+As expected, the latency story is an inverse of throughput,
+with 99th percentile being noisiest and showing anomalies in some Rust runs.
+The biggest (geometric) difference is in the 8--16 connections range where Kotlin frameworks already
+saturate available CPU, but Rust does not, yet.
+
+The latency "cutoff" at the very right of the graph corresponds with the change of mode:
+from saturation-induced latency to appearance of error responses.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/max_mem_usage.svg" />
+
+Here we measure high-water mark memory usage of the container from its start till the end of each
+benchmark step, as reported by Docker (i.e. not just momentary memory usage).
+
+The comparison between JVM-based and Rust frameworks is not representative here
+as allocated Java heap depends on configuration in addition to actual usage.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/max_mem_usage_per_requests.svg" />
+
+The same metric as above, but divided by the number of successful requests per second;
+which gives an unconventional unit of *megabyte-seconds per request*.
+
+It is normal that the values for low connections counts are higher
+because the base memory footprint of each framework is spread out across fewer requests.
+
+You may notice a slight difference in shape between Http4k and the 2 async frameworks:
+Http4k gets down to ~0.05 MiB⋅s and stays there, while the others go up again as latencies increase.
+If you have an explanation for this, please share it in a discussion.
```

Comment:
It's clear that the cause is the same for this and for the already [mentioned mystery](https://github.com/strohel/strohel.github.io/pull/2/files?short_path=7df25e1#diff-c1100ea69a8c13d3d16cf9875c68c6a2R328).

Might be worth noting in the text. .)

Reply:
Actually, I think this is not caused by
> The behaviour of Http4k in extreme connection counts is *opposite*, reaching its peak *successful* req/s of ~3500, despite the presence of non-zero error ratio.

...because it happens more early and in "normal" connection counts zone 32 to 256. Actix goes from ~0.0011 MiB⋅s to  ~0.0037, while Http4k stays still constant (albeit much higher).

But I do acknowledge that Http4k req/s, CPU efficiency and CPU usage are tightly related. I'll reword all the "mystery" sentences including this one, let's see.